File: Python/3. Longest Substring Without Repeating Characters.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        dic = {}
        result = 0
        length = 0
        left = 0
        for i, j in enumerate(s):
            if j not in dic:
                dic[j] = i
            else:
                if dic[j] >= left:
                    left = dic[j] + 1
                dic[j] = i
            length = len(s[left:i + 1])
            # print(length)
            # print(s[left:i+1])
            if length > result:
                result = length

        return result

File: Javascript/3. Longest Substring Without Repeating Characters.js
/**
 * @param {string} s
 * @return {number}
 */
var lengthOfLongestSubstring = function(s) {
    let maps = new Map()
    let result = 0
    let left = 0

    for (let i = 0; i < s.length; i++) {
        if (maps.has(s[i]) && left <= maps.get(s[i])) {
            left = maps.get(s[i]) + 1 
        }
        const length = i - left + 1
        result = Math.max(result, length)
        maps.set(s[i], i)
    }
    return result
};
